python/location/lbs/activity/audioset/dataprocessing/audio_processing.py

```diff
@@ -0,0 +1,262 @@
+  # Extracts the following features: chroma_stft, chroma_cqt, chroma_cens, melspectogram,
+  #                                  mfcc, rms, spectral_centroid, spectral_bandwidth,
+  #                                  spectral_contrast, spectral_flatness, spectral_rolloff,
+  #                                  poly_features, tonnetz, zero_crossing_rate
+  for video_id in downloaded_audio:
+    extract_features(video_id, src_dir)
+
+
+def extract_features(video_id, src_dir):
+  # Extract features from specified audio segments from given video_id
+  # input: str src_dir
+  # output: None
+
+  # Extract the following features from each downloaded 10 second audio segment using librosa
+  # Extracts the following features: chroma_stft, chroma_cqt, chroma_cens, melspectogram,
+  #                                  mfcc, rms, spectral_centroid, spectral_bandwidth,
+  #                                  spectral_contrast, spectral_flatness, spectral_rolloff,
+  #                                  poly_features, tonnetz, zero_crossing_rate
+  f_dict = {}
+  prefix = "sliced_"
+  suffix = ".wav"
+  # extract chroma_stft
+  y, sr = librosa.load(src_dir + "/" + prefix + video_id + suffix)
+  f_dict["chroma_stft"] = librosa.feature.chroma_stft(y, sr)
+  # extract chroma_cqt
+  f_dict["chroma_cqt"] = librosa.feature.chroma_cqt(y, sr)
+  # extract chroma_cens
+  f_dict["chroma_cens"] = librosa.feature.chroma_cens(y, sr)
+  # extract melspectrogram
+  f_dict["melspectrogram"] = librosa.feature.melspectrogram(y, sr)
+  # extract mfcc
+  f_dict["mfcc"] = librosa.feature.mfcc(y, sr)
+  # extract rms
+  f_dict["rms"] = librosa.feature.rms(y)
+  # extract spectral_centroid
+  f_dict["spectral_centroid"] = librosa.feature.spectral_centroid(y, sr)
+  # extract spectral_bandwidth
+  f_dict["spectral_bandwidth"] = librosa.feature.spectral_bandwidth(y, sr)
+  # extract spectral_contrast
+  f_dict["spectral_contrast"] = librosa.feature.spectral_contrast(y, sr)
+  # extract spectral_flatness
+  f_dict["spectral_flatness"] = librosa.feature.spectral_flatness(y)
+  # extract spectral_rolloff
+  f_dict["spectral_rolloff"] = librosa.feature.spectral_rolloff(y, sr)
+  # extract poly_features
+  f_dict["poly_features"] = librosa.feature.poly_features(y, sr)
+  # extract tonnetz
+  f_dict["tonnetz"] = librosa.feature.tonnetz(y, sr)
+  # extract zero_crossing_rate
+  f_dict["zero_crossing_rate"] = librosa.feature.zero_crossing_rate(y, sr)
+  feature_dict[video_id] = f_dict
+  print("extracted features")
+
+
+def create_csv_from_list():
+  # Creates a csv file to input into TensorFlow with the following structure: is_gunshot, video_id, feature list
+  # input: None
+  # output: None
+  with open("tfdataset.csv", 'w', newline='') as csvfile:
+    csv_writer = csv.writer(csvfile, delimiter=':', quoting=csv.QUOTE_MINIMAL)
+    csv_writer.writerow(
+      ['label', 'video_id', 'chroma_stft', 'chroma_cqt', 'chroma_cens', 'melspectrogram', 'mfcc', 'rms',
```

Comment:
share the same column order with write_row to make it less error-prone

Reply:
Could you elaborate on this? I'm not sure exactly what you mean.

Reply:
i think you're writing the header line here but the actual content in write_row function may be written in a different order if you don't pay attention. say you add a new feature "mfcc2" after mfcc but then you have to remember to modify here and write_row, which isn't a good sign.